### ISSUE TITLE
Order of calling close_spider in pipelines

### ISSUE THREAD

**Issue by debosmit** [scrapy/scrapy#2645]
`process_item` is called based on the order of the pipeline classes mentioned in [`ITEM_PIPELINES`](https://doc.scrapy.org/en/latest/topics/settings.html#std:setting-ITEM_PIPELINES) setting. But, `close_spider` follows the exact opposite order with the `close_spider` of the last pipeline getting called first [[link](https://github.com/scrapy/scrapy/blob/129421c7e31b89b9b0f9c5f7d8ae59e47df36091/scrapy/middleware.py#L64)].

I can't think of a good reason for this. Should I submit a PR?

**Comment by kmike** [comment]
I think the existing order makes sense. Let's say there is a pipeline A with order 100, and pipeline B with order 200. Call order is the following in this case:

1. A.open_spider
2. B.open_spider
3. A.process_item
4. B.process_item
5. B.close_spider
6. A.close_spider

It means pipeline B can depend on a pipeline A - when B.open_spider or B.close_spider are called A exists, and when B.process_item is called it gets result processed by A.

Does this ordering cause problems for you?

**Comment by debosmit** [comment]
@kmike Sure, I totally see the benefits in the `open_spider` and `process_item` cases. I am just not sure what benefit we get from calling `B.close_spider` *before* `A.close_spider`. Isn't it more intuitive for all the functions to be serial? For example, a call order like,

    1. A.open_spider
    2. B.open_spider
    3. A.process_item
    4. B.process_item
    5. A.close_spider
    6. B.close_spider

I stumbled upon this in a usage where, my `*pipeline A*` posts to ES and `*pipeline B*` then works on the indices touched by `*pipeline A*`. Say, `*pipeline A*` stores a buffer of items, that is completely flushed on close spider, then if `*pipeline B*`'s close_spider is called before `*pipeline A*`'s close_spider, the datastore has not yet been updated with all the data and it appears like the overall spider pipeline is behaving like a bi-directional valve. Given that an ordering for pipelines exists, I think there should be a way to support both orderings at the least.

Can you mention a use-case where you would have to have pipelines structured such that B would *require* being closed before A?

**Comment by kmike** [comment]
No, I haven't had a use case myself. I guess the problem is that there are two possible way of interaction:

1. middleware A keeps some resource "open", and middleware B uses this resource. In this case current order makes sense.
2. middleware A provides something in its close_spider method, and middleware B uses the result. In this case current order doesn't make sense.

I can't think of an use case for (1) too, but maybe I'm not creative enough. There is a lot of middlewares in a wild, so it looks like a rather big backwards incompatible change. But now I agree that (2) seems to make more sense.

What do you mean by supporting both orderings? Are you thinking of a setting, or something like that?

As a side note, I'm not sure I understand your example :) Why is it pipeline A who maintains a buffer?
If `*pipeline B*` adds something to an item then it makes sense for the last pipeline to have a buffer - it could be a pipeline B or even another pipeline C. Otherwise (if it is pipeline A which has a buffer) buffer may have outdated items which are dropped or modified but subsequent middlewares.

**Comment by debosmit** [comment]
I made a bad choice in wording. For my use-case, `*pipeline A*` post items to a ES in batches, hence it needs to hold the items till it has enough of them to make a batch request. Once, all the items have been batch-posted by `*pipeline A*`, `*pipeline B*` does some stuff based on the ES indices affected by `*pipeline A*`'s actions.

I totally agree with you about the backward compatibility point. I was thinking of some kind of setting that would allow the user to choose either having `close_spider` called in reverse order (as it is now), or else, have it called in the same order as that of the pipelines -- a simple boolean setting would suffice for this.
I would love to hear your thoughts and ideas about this.

**Comment by wanaryytel** [comment]
Whoa, this surprised (bit me in the ass I mean) as well. +1 to debosmit that this should be at least configurable.

**Comment by wanaryytel** [comment]
Submitted a PR for this: https://github.com/scrapy/scrapy/pull/3440

**Comment by Gallaecio** [comment]
I’m unsure about this change, but if we want to consider changing the default behavior (regardless of whether or not we allow overriding through a setting), Scrapy 2.0 would probably be the right release to do this.

**Comment by AdrianAtZyte** [comment]
> This is moot now. `open_spider`/`close_spider` for pipelines are no longer called serially, they all run concurrently (`ItemPipelineManager._process_parallel`), unlike `process_item`, which still goes through pipelines in order. So the order these are called in does not give any pipeline a guarantee about another one's state. Closing.
